### PR TITLE
Handle Symbol.toPrimitive in the BigDecimal class

### PR DESCRIPTION
**Pull request by d-mon-**
Hi!

I wanted to include a small improvement which helped us a lot during our migration to BigDecimal.
To do so, I extended the BigDecimal class to include `Symbol.toPrimitive` in order to mitigate human error by throwing an exception at runtime if a BigDecimal instance is used in arithmetic operations or concatenations (and more, see tests file). 

A few common issues for example: 
- `Big("5") - Big("10.0000000000000005")` => `-5` instead of `-5.0000000000000005`
- `Big("1e-50") > Big("10")` => `true` instead of `false` (due to string comparison)
- `Big(50) + Big(10)` => `"5010"` instead of `60`

I also noticed [a small mistake](5541ccc081f4522778748dd140f0a1d820318699) I made in one of my previous pull request and fixed it here.

### CODE REVIEW

File: src/bigdecimal.ts [hunk 1]
```diff
@@ -4350,6 +4350,37 @@ export class BigDecimal {
...
         }
     }
+
+    /**
+     * Function called in case BigDecimal is implicitly converted to a primitive value.
+     * The function will throw when the code try to convert the BigDecimal value in the following cases:
+     * - Using a greater or lower comparison operators (>, <, <=, >=)
+     * - Using loose equality operators against number, string, symbol or bigint (==, !=)
+     * - Using an arithmetic operators (+, -, /, *, %, **, --, ++)
+     * - Using bitwise operators (&, |, ^, ~, >>, >>>, <<)
```
Comment (srknzl) [inline src/bigdecimal.ts:4360]: Is there a place in the mozilla documentation where all the possible type coercions are listed? In each operator's documentation, the conversions are explained but I wonder if there is a list of operations that can do a call to toPrimitive.

File: src/bigdecimal.ts [hunk 2]
```diff
@@ -4350,6 +4350,37 @@ export class BigDecimal {
...
+     *
+     * Note:
+     * - In case of string templates, the function will return the value from {@link toString}.
+     * - It will throw when the plus (+) sign is used for addition or concatenation.
```
Comment (srknzl) [inline src/bigdecimal.ts:4364]: I think comparison operators will also make our function throw because 'number' hint is passed as I understand from the documentation. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Less_than 

Is there any other operation that can throw as well? Should we enumerate all of them in this Note? WDYT?

Also, concatenation will work like this, maybe we can remove the "it will throw when" statement? 

""+ Big("1.2123123")


I mean, because listing all the situations where we can throw seems hard to me. Even with just addition operator, only some cases throw exception. (e.g Big("1.2") + Big("2.1") )

File: test/convertToPrimitive.js [hunk 3]
```diff
@@ -0,0 +1,212 @@
+'use strict';
+const { Big } = require('../lib/bigdecimal.js');
+const chai = require('chai');
+chai.should();
+
+/* eslint-disable eqeqeq */
+describe('Convert to primitive test', function () {
+    it('should throw when calling JSON.stringify on BigDecimal', () => {
```
Comment (srknzl) [inline test/convertToPrimitive.js:8]: 🤔 Is this test related to converting to primitive?

File: test/convertToPrimitive.js [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        const testCase = () => JSON.stringify(Big(0));
+        testCase.should.throw('Do not know how to serialize a BigInt', undefined, 'expected JSON.stringify to throw');
+    });
+
+    it('should be able to convert to string (hint: string)', () => {
+        const value = Big(0);
+        const testCases = {
```
Comment (srknzl) [inline test/convertToPrimitive.js:15]: Maybe we can add `String(Big())` here as well?

File: test/convertToPrimitive.js [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+            '`${Big(0)}`': [() => `${value}`, '0'],
+        };
+
+        Object.entries(testCases).forEach(([key, [testCase, expected]]) => {
+            testCase().should.equals(expected, `expected '${key}' to be '${expected}'`);
+        });
+    });
+
+    it('should perform logical operators on BigDecimal without calling toPrimitive', () => {
```
Comment (srknzl) [inline test/convertToPrimitive.js:24]: if toPrimitive is not used with these, do we need this test? I mean, we are testing the Javascript's behaviour I think.

File: test/convertToPrimitive.js [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+                value ?? Symbol(1);
+            }
+        };
+
+        Object.entries(testCases).forEach(([key, testCase]) => {
+            testCase.should.not.throw(undefined, undefined, `expected '${key}' to not throw`);
+        });
+    });
+
+    it('should perform BigDecimal comparison without calling toPrimitive', () => {
```
Comment (srknzl) [inline test/convertToPrimitive.js:47]: similarly, do we need this test?

**Review by srknzl — CHANGES_REQUESTED** [review]
Looks good in general, have few comments

**Comment by d-mon-** [comment]
I forgot about this PR. I will close it for now as my focus has moved to a different project.
Thank you for the review and sorry for the delay.
